Implement basic layout and behaviour of new chat overlay

jai-x opened the pull request:
Part of https://github.com/ppy/osu/issues/17001

Reference design: https://www.figma.com/file/f8b2dHp9LJCMOqYP4mdrPZ/Client%2FChat

Provides initial implementation of new chat overlay in component
`ChatOverlayV2`. Contains only the basic functionality required for
a functioning chat overlay according to the new design with the intent
of added the rest of the functionality in subsequent PRs.

Backports existing tests for the current chat overlay except for ones
testing keyboard shortcuts (since they haven't been added) and tab
closing behaviour (since no tabs).

jai-x committed:
Implement basic layout and behaviour of new chat overlay

Provides initial implementation of new chat overlay in component
`ChatOverlayV2`. Contains only the basic functionality required for
a functioning chat overlay according to the new design with the intent
of added the rest of the functionality in subsequent PRs.

Backports existing tests for the current chat overlay except for ones
testing keyboard shortcuts (since they haven't been added) and tab
closing behaviour (since no tabs).

jai-x committed:
Remove blank line

jai-x committed:
Fix chat command test as reference equality checks on PM channels doesn't seem to to work

peppy commented:
Visual issues:

- Current channel needs to be white or something. This list looks very off as it stands:

![osu Game Tests 2022-05-02 at 06 25 49](https://user-images.githubusercontent.com/191335/166193194-0f8de0c7-80bb-4545-81a8-583742fed2fa.png)

- "in #xxx chatting" text is too long to fit (should not wrap). In the design it's aligned with the position the message would be once posted, so I think you should align it as per design:

![osu Game Tests 2022-05-02 at 06 27 42](https://user-images.githubusercontent.com/191335/166193344-5f224ee5-fa47-4b1e-a2e6-820fd5cf7459.png)

- On switching channels the textbox should be focused (also the search box in "Add more channels")

- Not sure "Add More Channels" should be title case, reads very weirdly.

- Day separator is not legible. There's a new version in the design - is that planned to be updated separately?

![osu Game Tests 2022-05-02 at 06 29 44](https://user-images.githubusercontent.com/191335/166193520-41cc4e43-3cf6-447a-9f14-33cbb2fdf329.png)

peppy reviewed (changes requested):
Haven't gone through the code in detail, just a preliminary visual/behavioural pass.

jai-x commented:
Addressing the points:

* Having the channel text be white indicates that the channel has some unread messages.
I will change it to white for now but the unread state design would need an update to differentiate it.
* Having the "in #xxx chatting" text not align was an oversight will fix that. However, it is noted in the figma design that this text is supposed to wrap when too long. Will change to truncate for now.
* Day separator not being updated was an oversight. I just used the base `DrawableChannel `implementation but I will subclass to use the expected design as done in other places in the codebase.

jai-x committed:
Ensure "chatting in..." text is aligned with chat message

jai-x committed:
Don't wrap "chatting in.." text in `ChatTextBar`

jai-x committed:
Change "Add more channels" to sentence case in "ChannelList"

jai-x committed:
Show selected channel text as white in `ChannelListItem`

jai-x commented:
I've addressed the visual changes aside from the day separator issue.
It seem that modifying that requires some awkward subclassing at which point it seems to better to do in it's own PR or to refactor the base implementation to support better customisation.

peppy commented on osu.Game/Overlays/Chat/ChannelList/ChannelListSelector.cs:
```diff
@@ -50,7 +50,7 @@ private void load(OverlayColourProvider colourProvider)
                         Anchor = Anchor.CentreLeft,
                         Origin = Anchor.CentreLeft,
-                        Text = "Add More Channels",
+                        Text = "Add more channels",
```
Why is this class not a `ChannelListItem`? It's reimplementing much of the same thing. It's also not getting the new selected state change.

jai-x replied:
Having the selector be a separate component was the easiest way to implement it at the time in #17833 but it does seem that it should be refactored to inherit from a base `ChannelListItem`.
Unsure if it's worth including that in this PR or should be updated separately.

bdach replied:
I mean if it were up to me then this diff would be half the current size, although I do realise that half of the diff is test backporting. I'd try to make the change here and see how large it gets, maybe there's a chance it'll be easy enough to review...

jai-x replied:
I had a go at doing this but I felt it would make a mess of this PR including it here so for now I've just updated the `ChannelListSelector` to turn the text white on select same as the items.

peppy commented:
Textbox focus doesn't seem addressed yet, did you take a look at that?

Also, I think the text next to the textbox needs a touch of padding:

![osu Game Tests 2022-05-03 at 07 33 54](https://user-images.githubusercontent.com/191335/166418384-f1b0c1c9-1fc9-48fe-bf1b-e4aee4d1d2a2.png)

jai-x committed:
Ensure `ChannelListSelector` text also turns white when selected

jai-x committed:
Add padding to text in `ChatTextBar` to separate it from the textbox

jai-x committed:
Ensure `ChatTextBox` takes/leaves focus on chat overlay pop in/out

peppy commented:
For textbox focus I think you can draw from the existing chat overlay's implementation, which also ensures it holds focus when you click away:


https://user-images.githubusercontent.com/191335/166676514-492abde2-acc9-494e-b2f0-7996dfe796d9.mp4


https://github.com/ppy/osu/blob/5e5c8e78a6538b4bcda713375f59b1b2828ec9fa/osu.Game/Overlays/ChatOverlay.cs#L218-L225

peppy committed:
Tidy up state variables

peppy reviewed (changes requested):
Focus still needs a bit of attention, as proposed.

jai-x committed:
Ensure focus is directed to `ChatTextBox` from `ChatOverlay` and add tests

peppy committed:
Merge branch 'master' into new-chat-overlay

peppy committed:
Reset all test data before each test method to avoid channels stuck in joined state

peppy committed:
Replace `HandlePositionalInput` override with simple hover effect

peppy committed:
Remove `!` usage (also seems to fix height saving/loading)

peppy committed:
Adjust value change variables to avoid `e` usage

peppy committed:
Tidy up initialisation

peppy commented:
@jai-x I've done a bit of tidying up, but looks pretty good for an initial effort apart from that. Please check the changes I made and make sure you are okay with them.

jai-x committed:
Ensure `RelativePositionAxes` is set in BDL for animations to work

jai-x committed:
Add tests for chat height saving/loading

jai-x commented:
I've had to re-add the `RelativePositionAxes` back into the `BDL` as it's required for the pop in/out animations to work correctly. Additionally, I've added some tests to verify the chat height load/save behaviour.

peppy commented:
New test doesn't work:

```csharp
System.NullReferenceException : Object reference not set to an instance of an object.
   at osu.Game.Tests.Visual.Online.TestSceneChatOverlayV2.TestChatHeight() in /Users/dean/Projects/osu/osu.Game.Tests/Visual/Online/TestSceneChatOverlayV2.cs:line 136

```

jai-x committed:
Fix chat height saving/loading test

peppy committed:
Merge branch 'master' into new-chat-overlay
